src/protocol/sort/secureapplyinv.rs

```diff
@@ -0,0 +1,171 @@
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::fabric::Network,
+    protocol::{
+        context::ProtocolContext,
+        reveal::reveal,
+        sort::ApplyInvStep::{RevealPermutation, ShuffleInputs, ShufflePermutation},
+        RecordId,
+    },
+    secret_sharing::Replicated,
+};
+use embed_doc_image::embed_doc_image;
+use futures::future::try_join_all;
+use permutation::Permutation;
+
+use super::{
+    apply::apply,
+    shuffle::{generate_random_permutation, Shuffle},
+};
+use tokio::try_join;
+
+/// This is APPLYINV(Algorithm 4) described in <https://eprint.iacr.org/2019/695.pdf>.
+/// This is a protocol that applies the inverse of a shared permutation to a shared-vector representation.
+/// Input: Each helpers know their own secret shares of input and permutation
+/// Output: At the end of the protocol, all helpers receive inputs after the permutation is applied
+#[derive(Debug)]
+#[embed_doc_image("secureapplyinv", "images/sort/secureapplyinv.png")]
+pub struct SecureApplyInv<'a, F> {
+    input: &'a mut Vec<Replicated<F>>,
+    permutation: &'a mut Vec<Replicated<F>>,
+}
+
+impl<'a, F: Field> SecureApplyInv<'a, F> {
+    #[allow(dead_code)]
+    pub fn new(input: &'a mut Vec<Replicated<F>>, permutation: &'a mut Vec<Replicated<F>>) -> Self {
+        Self { input, permutation }
+    }
+    /// This algorithm applies a permutation on the inputs. The permutation is not available in clear and
+    /// we do not want to reveal the actual permutation to any of the helpers while applying it.
+    /// Steps
+    /// ![Secure Apply Inv steps][secureapplyinv]
+    /// 1. Generate a permutation
+    /// 2. Secret shared permutation is shuffled with a permutation
+    /// 3. Secret shared value is shuffled using the same permutation as 1
+    /// 4. The permutation is revealed
+    /// 5. All helpers call `apply` to apply the permutation locally.
+    #[allow(dead_code)]
+    pub async fn execute<N: Network>(
+        &mut self,
+        ctx: &ProtocolContext<'_, N>,
+    ) -> Result<(), BoxError> {
+        // 1. Generate permutation
+        let permutations = generate_random_permutation(self.input.len(), &ctx.prss());
```

Comment:
Is it just one permutation, or multiple? The function name is singular but the variable is plural.

Reply:
Actually, I think the confusing thing about this is actually a much deeper issue.

I don't think it makes sense for the caller of the `Shuffle` protocol to call into the `shuffle` module and invoke the `generate_random_permutation` function themselves. This should be a private function invoked internally by the `Shuffle` protocol. 

I'd recommend changing the implementation of the `Shuffle.execute()` method to just internally `narrow` the context that is passed it, to then generate two out of the three random permutations, and then apply them.

Reply:
Oh wait... nevermind... but then how do we apply the same exact random permutation to both the inputs and the shuffle_permutation... I see why you did it this way...

OK, let's at least make it somehow more clear (via the name?) that you're generating two out of the three permutations that will eventually be applied. This is (in some sense) a secret-share of the full permutation. Perhaps call it: "random_permutation_share".

Reply:
Yes, the reason is exactly that we want same permutation to be used for both input and permutation. 
It generates two permutations per helper (left and right) that's why has a "s" in the name. Both the permutations are used while executing the shuffle protocol.
For naming, worried "share" can cause more confusion, so renaming it to "generate_random_permutations_using_prss" to reflect the role of prss hence common permutations across helpers?

Reply:
That should help. Let’s pick a different local variable name as well. Something like “two_of_three_permutations”